examples/mauna-loa/script.jl

```diff
@@ -0,0 +1,301 @@
+    return -lml
+end
+#md nothing #hide
+
+# !!! note "Work-in-progress"
+#     In the future, we are planning to provide the `optimize_loss` utility
+#     function as part of JuliaGaussianProcesses -- for now, we just define it
+#     inline.
+#
+# The L-BFGS parameters were chosen because they seem to work well empirically.
+# You could also try with the defaults.
+
+default_optimizer = LBFGS(;
+    alphaguess=Optim.LineSearches.InitialStatic(; scaled=true),
+    linesearch=Optim.LineSearches.BackTracking(),
+)
+
+function optimize_loss(loss, θ_init; optimizer=default_optimizer, maxiter=1_000)
+    options = Optim.Options(; iterations=maxiter, show_trace=true)
+
+    θ_flat_init, unflatten = ParameterHandling.value_flatten(θ_init)
+    loss_packed = loss ∘ unflatten
+
+    ## https://julianlsolvers.github.io/Optim.jl/stable/#user/tipsandtricks/#avoid-repeating-computations
+    function fg!(F, G, x)
+        if F !== nothing && G !== nothing
+            val, grad = Zygote.withgradient(loss_packed, x)
+            G .= only(grad)
+            return val
+        elseif G !== nothing
+            grad = Zygote.gradient(loss_packed, x)
+            G .= only(grad)
+            return nothing
+        elseif F !== nothing
+            return loss_packed(x)
+        end
+    end
+
+    result = optimize(Optim.only_fg!(fg!), θ_flat_init, optimizer, options; inplace=false)
+
+    return unflatten(result.minimizer), result
+end
+#md nothing #hide
+
+# We now run the optimization:
+
+θ_opt, result = optimize_loss(loss, θ_init)
+result
+
+# The final value of the log marginal likelihood is:
+
+-result.minimum
+
+# !!! warning
+#     To avoid bad local optima, we could (and should) have carried out several
+#     random restarts with different initial values for the hyperparameters,
+#     and then picked the result with the highest marginal likelihood. We omit
+#     this for simplicity. For more details on how to fit GPs in practice,
+#     check out [A Practical Guide to Gaussian
+#     Processes](https://tinyurl.com/guide2gp).
+#
+# Let's construct the posterior GP with the optimized hyperparameters:
+
+fpost_opt = build_posterior_gp(ParameterHandling.value(θ_opt))
+#md nothing #hide
+
+# This is the kernel with the point-estimated hyperparameters:
+
+fpost_opt.prior.kernel
+
+# Let's print the optimized values of the hyperparameters in a more helpful format:
+
+# !!! note "Work-in-progress"
+#     This is another utility function we would eventually like to move out of this notebook:
+
+using Printf
+
+show_params(nt::Union{Dict,NamedTuple}) = String(take!(show_params(IOBuffer(), nt)))
+function show_params(io, nt::Union{Dict,NamedTuple}, indent::Int=0)
+    for (s, v) in pairs(nt)
+        if v isa Union{Dict,NamedTuple}
+            println(io, " "^indent, s, ":")
+            show_params(io, v, indent + 4)
+        else
+            println(io, " "^indent, s, " = ", @sprintf("%.3f", v))
+        end
```

Comment:
There are some avoidable string concatenations left (`" "^indent` and `@sprintf` will allocate strings with whitespace and not print them directly to `io`) that one can get rid of. To reduce the amount of code I moved the identical operations out of the `if` block:
```suggestion
        for _ in 1:indent
            print(io, " ")
        end
        print(io, s)
        if v isa Union{Dict,NamedTuple}
            println(io, ":")
            show_params(io, v, indent + 4)
        else
            println(io, " = ")
            @printf(io, "%.3f", v)
        end
```

Reply:
Yes, this would be marginally faster, but I found the version as it is in the script now rather more readable, and the impact on performance is negligible.

Reply:
It's good to write performant code when that's easy, but speed doesn't trump everything else, particularly not when it's a few percent shaved off a function that's nowhere near the bottleneck and already running in microseconds not milliseconds..

Reply:
Sure, this is not completely satisfying for a tutorial (even though I think the more verbose version is more readable :shrug:) - but IMO neither of the two versions is. But I assumed this will be moved somewhere else soon, and hence I thought it would be better to not having to think about it and optimize it again when it's moved.

Reply:
Yes, this should get moved elsewhere so it's easy to re-use wherever! I'd still rather keep it like this though.